src/library/scala/collection/immutable/List.scala

```diff
@@ -560,6 +567,7 @@ sealed abstract class List[+A]
 
 final case class :: [+A](override val head: A, private[scala] var next: List[A @uncheckedVariance]) // sound because `next` is used only locally
   extends List[A] {
+  ScalaRunTime.releaseFence()
```

Comment:
Let's add a comment to the class that every time `next` is mutated and the resulting `::` instance is published, you should first call `releaseFence`. Maybe, *maybe*, we can actually add a method `publishNext` or something to make this more clear.